### PR TITLE
Feature/56 select on map

### PR DESCRIPTION
**Pull request by Eirmas**
Implemented the select on map tool. It can now move groups of shapes around.

Demo:
Move demo

### CODE REVIEW

File: src/components/TheToolPanel.vue [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <template>
   <div class="btn-toggle-box ma-0 pa-0">
     <v-btn-toggle class="flex-column">
+      <tool-container :id="id" icon="fa-arrows-alt" toolname="move" :popout="false" />
```
Comment (bryansamuel1) [inline src/components/TheToolPanel.vue]: Can we use a different icon here than the arrows. This tool will also be used to select entities on the canvas so you won't necessarily be moving a shape all the time. It needs to be a select icon

File: src/tools/Move.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { Tool, Tracker } from '@/tools/Tool'
+import Konva from 'konva'
+import { CanvasElement } from '@/types/Canvas'
+import throttle from 'lodash.throttle'
+
+export default class FreeDraw implements Tool {
+  // eslint-disable-next-line
+  private group: Konva.Group
+  constructor (public readonly name: string,
+               public temporary: boolean) {
```
Comment (bryansamuel1) [inline src/tools/Move.ts]: is temporary applicable for this tool? Do we need to have a property if it won't be used?

File: src/tools/Move.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      canvasElement.id = this.group.attrs.id
+    }
+  }
+
+  // eslint-disable-next-line
+  mouseMoveAction = throttle((e: Konva.KonvaPointerEvent, canvasElement: CanvasElement, layer: Konva.Layer, socket: WebSocket): void => {
+    if (!canvasElement.hasMoved) {
+      canvasElement.hasMoved = true
+    }
+    if (canvasElement.id !== '') {
```
Comment (bryansamuel1) [inline src/tools/Move.ts]: `canvasElement.id` is better, or are we just checking that it isn't an empty string. What about if it is null or undefined?

File: src/tools/Move.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      const pos = { x: (e.evt.x - canvasElement.data[0]), y: (e.evt.y - canvasElement.data[1]) }
+      this.group.move(pos)
+      layer.batchDraw()
+      canvasElement.data = [e.evt.x, e.evt.y]
+    }
+  }, 0)
+
+  mouseUpAction = (e: Konva.KonvaPointerEvent, canvasElement: CanvasElement, _layer: Konva.Layer, socket: WebSocket): void => {
+    if (canvasElement.hasMoved && canvasElement.id !== '') {
```
Comment (bryansamuel1) [inline src/tools/Move.ts]: same again here for the canvasElement.id

File: src/tools/UndoRedo.ts [hunk 5]
```diff
@@ -5,19 +5,12 @@ export default class UndoRedo {
...
-          layerId: lastElement.layerId,
-          tracker: Tracker.UNDO,
-          data: lastElement.data,
-          jti: lastElement.jti,
-          tool: lastElement.tool,
-          change: true,
-          hasMoved: true
-        }
+        const copy = [...canvasElementsHistory]
+        copy.splice(copy.length - 1, 1)
```
Comment (bryansamuel1) [inline src/tools/UndoRedo.ts]: use `copy.pop()` to remove the last element instead

File: src/tools/UndoRedo.ts [hunk 6]
```diff
@@ -39,19 +32,10 @@ export default class UndoRedo {
...
-          tracker: Tracker.REDO,
-          data: lastElement.data,
-          jti: lastElement.jti,
-          tool: lastElement.tool,
-          change: true,
-          hasMoved: true
-        }
+      if (lastElement.tracker === Tracker.UNDO) {
+        const copy = [...canvasElementsHistory]
+        copy.splice(copy.length - 1, 1)
```
Comment (bryansamuel1) [inline src/tools/UndoRedo.ts]: `copy.pop()` to remove the last element instead

File: src/tools/UndoRedo.ts [hunk 7]
```diff
@@ -66,6 +50,8 @@ export default class UndoRedo {
           copy.splice((copy.length - 1), 1)
```
Comment (bryansamuel1) [inline src/tools/UndoRedo.ts]: and against with the pop()

File: src/tools/shapes/CircleCreator.ts [hunk 8]
```diff
@@ -81,6 +81,10 @@ export default class CircleCreator implements Shape {
     layer.batchDraw()
   }
 
+  getGroup = (): Konva.Group => {
+    return this.group
```
Comment (bryansamuel1) [inline src/tools/shapes/CircleCreator.ts]: could you add this on one line?
`getGroup = (): Konva.Group => this.group`

File: src/tools/shapes/FreedrawCreator.ts [hunk 9]
```diff
@@ -39,4 +39,8 @@ export default class FreedrawCreator implements Shape {
       id: canvasElement.id
     })
   }
+
+  getGroup = (): Konva.Group => {
+    return this.group
```
Comment (bryansamuel1) [inline src/tools/shapes/FreedrawCreator.ts]: same here

File: src/tools/shapes/LineCreator.ts [hunk 10]
```diff
@@ -151,6 +151,10 @@ export default class LineCreator implements Shape {
     layer.batchDraw()
   }
 
+  getGroup = (): Konva.Group => {
+    return this.group
```
Comment (bryansamuel1) [inline src/tools/shapes/LineCreator.ts]: and here

File: src/tools/shapes/PingCreator.ts [hunk 11]
```diff
@@ -47,4 +47,8 @@ export default class PingCreator implements Shape {
       layer.batchDraw()
     }, this.period)
   }
+
+  getGroup = (): Konva.Group => {
+    return this.group
```
Comment (bryansamuel1) [inline src/tools/shapes/PingCreator.ts]: and here

File: src/tools/shapes/RulerCreator.ts [hunk 12]
```diff
@@ -122,6 +122,10 @@ export default class RulerCreator implements Shape {
     layer.batchDraw()
   }
 
+  getGroup = (): Konva.Group => {
+    return this.group
```
Comment (bryansamuel1) [inline src/tools/shapes/RulerCreator.ts]: and here

File: src/tools/shapes/TextCreator.ts [hunk 13]
```diff
@@ -94,4 +94,8 @@ export default class TextCreator implements Shape {
     this.textArea.focus()
     return this.textArea
   }
+
+  getGroup = (): Konva.Group => {
+    return this.group
```
Comment (bryansamuel1) [inline src/tools/shapes/TextCreator.ts]: and here

File: src/components/TheToolPanel.vue (hunk 1 above)
Comment (Eirmas) [reply src/components/TheToolPanel.vue]: I think Rites is creating the icon atm so will leave this at that

File: src/tools/Move.ts (hunk 3 above)
Comment (Eirmas) [reply src/tools/Move.ts]: Has to look for empty string

File: src/tools/Move.ts (hunk 4 above)
Comment (Eirmas) [reply src/tools/Move.ts]: Has to look for empty string

File: src/tools/Move.ts (hunk 2 above)
Comment (Eirmas) [reply src/tools/Move.ts]: I made it a standard for all tools to include the name and the temporary variable in the constructor

File: src/tools/shapes/CircleCreator.ts [hunk 14]
```diff
@@ -81,12 +81,10 @@ export default class CircleCreator implements Shape {
...
   }
 
-  getGroup = (): Konva.Group => {
-    return this.group
-  }
-
   // eslint-disable-next-line
   [key: string]: any;
+
+  getGroup = (): Konva.Group => this.group
```
Comment (bryansamuel1) [inline src/tools/shapes/CircleCreator.ts]: Could we do anything about the repetitive code with getGroup? 
Could each shape extend a class that has this and other common methods in it? So each shape then inherits/extends these method and we only have one place to update?

**Pull request merged**